ding_unilogin/src/OAuth2/Client/Provider/StilUniloginProvider.php

```diff
@@ -0,0 +1,142 @@
+      parent::getRequiredOptions(),
+      [
+        'urlTokenIntrospect',
+        'urlEndSession',
+      ]
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function getConfigurableOptions() {
+    return array_merge(
+      parent::getConfigurableOptions(),
+      [
+        'urlTokenIntrospect',
+        'urlEndSession',
+      ]
+    );
+  }
+
+  /**
+   * Introspect token, i.e. get user info.
+   *
+   * @param \League\OAuth2\Client\Token\AccessToken $accessToken
+   *   The access token.
+   *
+   * @return array
+   *   The user info.
+   *
+   * @see https://viden.stil.dk/display/OFFSKOLELOGIN/Implementering+af+tjeneste#Implementeringaftjeneste-1.3.3ModtagbrugerinformationermedAccessToken
+   */
+  public function introspectToken(AccessToken $accessToken): array {
+    $request = $this->getRequest('POST',
+      $this->urlTokenIntrospect,
+      [
+        'headers' => [
+          'content-type' => 'application/x-www-form-urlencoded',
+        ],
+        'body' => $this->buildQueryString([
+          'token' => $accessToken->getToken(),
+          'client_id' => $this->clientId,
+          'client_secret' => $this->clientSecret,
+        ]),
+      ]
+    );
+    $response = $this->getResponse($request);
+
+    try {
+      $content = $response->getBody()->getContents();
+      return json_decode($content, TRUE, 512, JSON_THROW_ON_ERROR);
+    } catch (JsonException $jsonException) {
+      throw new StilUniloginException('Cannot parse token', $jsonException->getCode(), $jsonException);
+    } catch (\Exception $exception) {
+      throw new StilUniloginException('Cannot introspect token', $exception->getCode(), $exception);
+    }
+  }
+
+  /**
+   * End Unilogin session.
+   *
+   * @param string $postLogoutRedirectUri
+   *   The post logout redirect URI.
+   * @param string $idTokenHint
+   *   The ID token hint.
+   *
+   * @see https://viden.stil.dk/display/OFFSKOLELOGIN/Implementering+af+tjeneste#Implementeringaftjeneste-1.5LogudmedOIDC
+   */
+  public function endSession(string $postLogoutRedirectUri, string $idTokenHint) {
+    $parameters = [
+      'post_logout_redirect_uri' => $postLogoutRedirectUri,
+      'id_token_hint' => $idTokenHint,
+    ];
+
+    // "Logud håndteres af Unilogin brokeren, og kræver derfor kun et get request med redirectUri som parameter. Der arbejdes på at kunne understøtte Frontchannel og Backchannel Logout." (cf. https://viden.stil.dk/display/OFFSKOLELOGIN/Implementering+af+tjeneste#Implementeringaftjeneste-1.5LogudmedOIDC_
+    // drupal_goto(url($this->urlEndSession, ['query' => $parameters]));
+    // Send the GET request and swallow select exceptions.
+    try {
+      $request = $this->getRequest('GET', url($this->urlEndSession, ['query' => $parameters]));
+      $this->getResponse($request);
+    }
+    catch (ConnectException $connectException) {
+      // ConnectException may be thrown during local development if the redirect
+      // URI cannot be accessed from PHP.
```

Comment:
Should we log this to watchdog, if it happens in prod, to make live debug easier?

Reply:
Catching of `ConnectException` is only added for local development, so it should never happen in production. And if it does, we really don't care – if the user's session is not ended we'll get another exception which is caught and logged elsewhere.